### PR TITLE
Templatized the MultiDeviceObject, -Resource and -Pool classes.

### PR DESCRIPTION
**Pull request by jhmueller-huawei**
## What does this PR do?

This PR is part of https://github.com/o3de/sig-graphics-audio/pull/137 (original proposal including discussion in https://github.com/o3de/sig-graphics-audio/issues/120, there referred to as (part of) commit 3 in the Planned git history sub-section) and introduces templated versions of 3 abstract base classes to simplify `MultiDeviceObject` based classes as discussed in https://github.com/o3de/o3de/pull/16262#discussion_r1271205299

## Planned PRs
This commit is one in a series of PRs to come:

| Name | Link | Status |
| --- | --- | --- |
|Preparation|[#16138](https://github.com/o3de/o3de/pull/16138)|merged|
|Base Resources|[#16202](https://github.com/o3de/o3de/pull/16202)|merged|
| Pipelines | [#16261](https://github.com/o3de/o3de/pull/16261) | merged |
|Independent Resources| [#16262](https://github.com/o3de/o3de/pull/16262) | merged|
| Templates | this PR | open |
| Buffers | [#16590](https://github.com/o3de/o3de/pull/16590) | open |
| Images | [#16591](https://github.com/o3de/o3de/pull/16591) | open |
|Indirect*|||
| Items | | |
| SRGs| | |
|RayTracing Resources| | |

## How was this PR tested?

Using the RHI tests including the already existing multi device tests.

### CODE REVIEW

**Review by moudgils — COMMENTED** [review]
Much better. Just a few minor points. Thanks.

File: Gems/Atom/RHI/Code/Include/Atom/RHI/MultiDeviceObject.h [hunk 1]
```diff
@@ -8,34 +8,59 @@
...
 #include <Atom/RHI/Device.h>
+#include <Atom/RHI/RHISystemInterface.h>
 
 namespace AZ::RHI
 {
     //! A variant of Object associated with a DeviceMask.
     //! In contrast to DeviceObject, which is device-specific and holds a strong reference to a specific device, 
     //! MultiDeviceObject only specifies on which device an object resides/operates, specified by a
     //! DeviceMask (1 bit per device).
+    template <class DeviceObject>
```
[the remark this answers is not in the document]
Comment (jhmueller-huawei) [reply Gems/Atom/RHI/Code/Include/Atom/RHI/MultiDeviceObject.h]: Changed it to `DeviceObjectClass`.

**Review by moudgils — APPROVED** [review]

**Comment by jhmueller-huawei** [comment]
Working on using these classes in the RPI, I just realized that having it templated is an issue because now I cannot use polymorphism anymore, i.e., having a pointer of a `MultiDeviceResource` pointing to an instance of a derived object. The only solution there would be to create another interface class that provides the pure virtual functions that are needed for the polymorphic use cases. I will add that in a later PR if you don't have any other suggestions or want it immediately in this one. The same is btw true for `ResourceView` which is also used with polymorphically, so we will have to introduce some lightweight form of MD resource views and an abstract base class for it.

**Comment by jhmueller-huawei** [comment]
> Now that I see the changes with all of the templates, I think this adds too much complexity in the usage (specially for handling MultiDeviceResources and MultiDevicePools). What about just moving the AZStd::unordered_map with the DeviceObject pointers, and add a protected templated function GetDeviceObject that returns the proper casted type?. This way we avoid using template in the MultiDeviceObject class (which is the part that creates the problems)

The `GetDeviceObject` method is pretty much exclusively used by the public interface and I would not have a templated version for that because it would allow you to call `GetDeviceObject<Image>()` on a `MultiDeviceBuffer` which would of course fail, but the nicer API (as we have it here) would not allow this in the first place. The alternative is to have a `GetDevice*()` method in every class like it is now without this PR (e.g. `GetDeviceBuffer()`), which does the cast (or uses the protected templated method).

What we need usually within the MD classes is not access to a single `DeviceObject`, but looping over all of them in the map. We could do this with a template function just like `IterateDevices` (which is used for initialization, i.e., filling the map the first time), instead of the for each loop we have with and without this patch, rewriting all these loops to a method call with a lambda.

Considering both of these points I would actually prefer to simply have a correctly typed map in each class. It's less, simpler and nicer code IMO. Or we go with this PR here (introducing interface classes for the cases where we need polymorphism as mentioned in the comment before).

As a last and quite important note: Please let me know asap what we should continue with since this has a considerable impact on everything that follows and we are starting to be really late with our multi-device project here and I fear if we cannot speed up the process, the project will be canceled soon.

**Comment by jhmueller-huawei** [comment]
Ok, I just ran into the issue that with the inheritance hierarchy of `ImagePool`, `StreamingImagePool` and `SwapChain` deriving from `ImagePoolBase`, this template solution doesn't work either: `MDImage` needs a pool as template parameter and the `MDImagePoolBase` doesn't work, because then the `m_deviceObjects` have the wrong type in the derived pool classes. We'd have to find a solution there as well and I can't even think of one, not even with an interface class…

I'm very much in favor of dropping any templated solutions here now and stick with what we have, i.e., a map per (final) class.

**Comment by moudgils** [comment]
> > > Now that I see the changes with all of the templates, I think this adds too much complexity in the usage (specially for handling MultiDeviceResources and MultiDevicePools). What about just moving the AZStd::unordered_map with the DeviceObject pointers, and add a protected templated function GetDeviceObject that returns the proper casted type?. This way we avoid using template in the MultiDeviceObject class (which is the part that creates the problems)
> > 
> > 
> > The `GetDeviceObject` method is pretty much exclusively used by the public interface and I would not have a templated version for that because it would allow you to call `GetDeviceObject<Image>()` on a `MultiDeviceBuffer` which would of course fail, but the nicer API (as we have it here) would not allow this in the first place. The alternative is to have a `GetDevice*()` method in every class like it is now without this PR (e.g. `GetDeviceBuffer()`), which does the cast (or uses the protected templated method).
> > What we need usually within the MD classes is not access to a single `DeviceObject`, but looping over all of them in the map. We could do this with a template function just like `IterateDevices` (which is used for initialization, i.e., filling the map the first time), instead of the for each loop we have with and without this patch, rewriting all these loops to a method call with a lambda.
> > Considering both of these points I would actually prefer to simply have a correctly typed map in each class. It's less, simpler and nicer code IMO. Or we go with this PR here (introducing interface classes for the cases where we need polymorphism as mentioned in the comment before).
> > As a last and quite important note: Please let me know asap what we should continue with since this has a considerable impact on everything that follows and we are starting to be really late with our multi-device project here and I fear if we cannot speed up the process, the project will be canceled soon.
> 
> I think is clear that the template approach is no good (to much issue with inheritance). My preference is doing the approach that you mentioned of adding the "GetDeviceObject" or "GetDeviceX" in each class that handles the proper casting. We can even use a macro for this similar to what we do with the AZ_CLASS_ALLOCATOR or AZ_RTTI macros. What do you think @moudgils ?

I prefer the approach outlined by Akio here. Just add a GetDeviceObject that handles the casting. Will that work?

**Comment by moudgils** [comment]
> > Ok, I've now created two alternatives in #16635 and #16636 - the first is using a map of `DeviceObject` and casts and in I don't think this turns out in as nice code as I would like to have it. You already suggested having a macro (which I didn't put in the first approach yet), so I thought that if we already have macros, we can use them to create the map as well, which is what I show in the second approach. While I don't particularly like macros, I think the second approach still leads to better code. What do you think?
> 
> Personally I prefer option #16635, but I'm okay with using option #16636

Same. I prefer https://github.com/o3de/o3de/pull/16635 as well. I have also start a AR run for that option. Thanks for implementing both to compare.

**Pull request closed without merging**